Recursive zk proofs da edits

anselld opened the pull request.

anselld commented:
Hi @Empiech007 and @hansieodendaal, please could you review my suggested edits. I've checked all the links as well.
Tony, for consistency between the labels in the figures and the text in the report, please could you change the spelling of 'Initialise' in Figures 1, 2 and 6 to 'Initialize'.
Please let me know if you have any questions or comments.
Thanks and regards, Denise.

Empiech007 commented:
Thank you @anselld ... will look into it.

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -1,729 +1,497 @@
+- [What is Recursive Proof Composition?](#what-is-recursive-proof-composition?)
 	- [Recursive Functions](#recursive-functions)
-	- [Recursion in Bulletproofs Inner-Product Proofs](#recursion-in-bulletproofs-inner-product-proof)
+	- [Recursion in Bulletproofs Inner-product Proofs](#recursion-in-bulletproofs-inner-product-proofs)
```
Please rather leave it as "- [Recursion in Bulletproofs Inner-product Proof](#recursion-in-bulletproofs-inner-product-proof)" to tally with the title of the report. i.e., keep it singular ... 'proof'.

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -1,729 +1,497 @@
+vectors using $\mathcal{O}(log(n))$ steps, and it’s the reason that Bulletproofs are compact" [[10]]. 
 
+No wonder the Halo creators also looked at the IPP in their amortization strategies, particularly taking advantage of its 
+recursive nature, . Close attention is therefore given to the IPP as described by Bootle et al. [[9]]. 
```
Note the extra "," in line 115 ... so change "nature,  ."  to  "nature."

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
Insert "of" between "number" and "multiplication" at the end of line 421.

Empiech007 replied:
In fact, make the last part of line 421 read as "in terms of the number of multiplications."

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
the last part of this line was meant to be ... "running from $k = log_2(n)$ down to $1$."

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
insert the article "the" between "of" and "number"

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -735,508 +503,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
-The amortized inner-product proof lends itself useful in implementations of any zero-knowledge proof that involves an inner-products especially in the Bulletproofs framework. 
-
-
+(d) The only case in which it is on par with the naive algorithm, is when all other algorithms are on par, and when for 
+the smallest cases $n = 4$ and  $n = 8$. 
```
Let us end the sentence as ... "and for the smallest cases when $n = 4$ and  $n = 8$."

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -735,508 +503,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
+- Each algorithm aims at reducing verification costs in terms of the number of multiplications. 
 
-- each algorithm aims at reducing verification costs in terms of the number of multiplications, 
+- Each algorithm takes advantage of the fact that each coefficient has common sub-product $2^{k-2}$ coefficients. 
```
Let us keep the end as it was originally ... "each coefficient has common sub-products with other $2^{k-2}$ coefficients,"

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -735,508 +503,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
+- Each algorithm takes advantage of the fact that each coefficient has common sub-product $2^{k-2}$ coefficients. 
 
-- each algorithm takes advantage of the fact that each coefficient has common sub-products with other  $2^{k-2}$ coefficients, 
+- The intermediate values of the coefficients are not used anywhere in the IPP. Only the final values make up all 
+$k$ factors $u_j^{b(i,j)}$. Hence their sub-products can be separately computed and only put together at the end of the 
+IPP's $k-$th round. 
```
Let us use this version: 
"The intermediate values of the coefficients are not used anywhere in the IPP. Only their final values made up of all 
$k$ factors, $u_j^{b(i,j)}$. Hence their sub-products can be separately computed and only put together at the end of the 
IPP's $k-$th round."

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -735,508 +503,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
-
-
+**Algorithm 4**: This algorithm is the same as Algorithm 3 throughout the IPP rounds. However, at the end of the IPP 
+rounds, the program gives preference to the formation all possible distinct quadruples. Larger-sized sub-products are then 
```
insert "of" between "formation" and "all". So that the sentence that starts in line 798 ends as ... 
"the program gives preference to the formation of all possible distinct quadruples."

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
````diff
@@ -735,508 +503,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
+```
 **Example 1 (Algorithm 1 or [A1])** 
 
+Let $n = 32$ so that $k = 5$. The coefficients $s_i$ for $i \in {0, 1, 2, ... , 15}$ are 32$ quintets: 
````
end the line as ... 
"are $32$ quintets:"

anselld commented:
Hi @Empiech007, thanks so much for the review and comments. I've updated my version accordingly, and will push the latest version. I'll let you know via Telegram, when I've done this. Best regards, Denise.

Empiech007 commented:
Hi @anselld, with these 3 changes and no further changes on your side, the report is approved (on my side). It appears @hansieodendaal still has to look at it.

hansieodendaal commented:
Thank you @Empiech007 and @anselld!
@anselld, please update changes as requested, commit and push, then will give it a quick once over.

anselld commented:
Changes updated as requested, @Empiech007 and @hansieodendaal. Please let me know if there is anything else. I also noticed that on the rendered version on the first (a), (b), (c) list (line 403), the (a), (b) and (c) should be on separate lines. I corrected this.

Empiech007 commented:
It's all good. Thanks @anselld 
We now await @hansieodendaal's approval.

hansieodendaal commented:
Presentation in mdbooks looks good, just some questions:

hansieodendaal commented:
Is this intentional?
![image](https://user-images.githubusercontent.com/39146854/82429683-9d194780-9a8c-11ea-895b-6820333de14e.png)

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -1,729 +1,499 @@
+As noted earlier, this amortization strategy reduces the verification costs by factor of $log(p)$. 
 
-### Verifiable Computation: Application 2
+### Verifiable Computation - Application 2
```
I think this content justifies more sub-level headings and not just boldface.

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -1,729 +1,499 @@
 
+### Verifiable Computation - Application 1
 
 **Delegating Inversion of Verifier's Challenges** 
```
This could be a sub-level heading

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -735,508 +505,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
+[14]: https://eprint.iacr.org/2019/1021.pdf "Recursive Proof Composition 
+without a Trusted Setup"
 
 ## Appendix A 
```
I think this content justifies more sub-level headings and not just boldface.

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
This formula renders incorrectly. The opening and closing `$$` should be on a separate line each time.

![image](https://user-images.githubusercontent.com/39146854/82430650-e9b15280-9a8d-11ea-91c2-0848ff262dfa.png)

hansieodendaal replied:
Try this:
```
$$
(s_{1} \cdot s_{n-1}  = u_1^2) \land (s_{2} \cdot s_{n-1} = u_2^2) \land (s_{4} \cdot s_{n-1}  = u_3^2) \land\ \dots \ \land (s_{2^{(k-1)}} \cdot s_{n-1}  = u_k^2) \land \\\\ 
(s_0 \cdot s_{(n-1) - 2^{(0)}}  = u_1^{-2}) \land (s_0 \cdot s_{(n-1) - 2^{(1)}}  = u_2^{-2}) \land\ \dots \ \land (s_0 \cdot s_{(n-1) - 2^{(k-1)}}  = u_k^{-2})  \land \\\\ 
( s_{r_1} \cdot s_{(n-1) - {r_1}}  = 1_{\mathbb{F}\_p} ) \land (s_{r_2} \cdot s_{(n-1) - {r_2}}  = 1_{\mathbb{F}\_p}) \land (s_{r_3} \cdot s_{(n-1) - {r_3}}  = 1_{\mathbb{F}\_p}) \land\ \dots \ \land (s_{r_k} \cdot s_{(n-1) - r_k}  = 1_{\mathbb{F}\_p})  
$$
```
It renders like this:
![image](https://user-images.githubusercontent.com/39146854/82433033-229ef680-9a91-11ea-81c6-2b3ef683eb6b.png)

hansieodendaal replied:
I replaced `\and` with `\land`, not sure if that is the intention, as `\and` is not recognized.

anselld replied:
@Empiech007, please check if this is okay.

Empiech007 replied:
I can't see the change. Hansie's version works perfectly, so simply copy this 

$$
(s_{1} \cdot s_{n-1}  = u_1^2) \land (s_{2} \cdot s_{n-1} = u_2^2) \land (s_{4} \cdot s_{n-1}  = u_3^2) \land\ \dots \ \land (s_{2^{(k-1)}} \cdot s_{n-1}  = u_k^2) \land \\\\ 
(s_0 \cdot s_{(n-1) - 2^{(0)}}  = u_1^{-2}) \land (s_0 \cdot s_{(n-1) - 2^{(1)}}  = u_2^{-2}) \land\ \dots \ \land (s_0 \cdot s_{(n-1) - 2^{(k-1)}}  = u_k^{-2})  \land \\\\ 
( s_{r_1} \cdot s_{(n-1) - {r_1}}  = 1_{\mathbb{F}\_p} ) \land (s_{r_2} \cdot s_{(n-1) - {r_2}}  = 1_{\mathbb{F}\_p}) \land (s_{r_3} \cdot s_{(n-1) - {r_3}}  = 1_{\mathbb{F}\_p}) \land\ \dots \ \land (s_{r_k} \cdot s_{(n-1) - r_k}  = 1_{\mathbb{F}\_p})  
$$ 

and replace the old version.

Empiech007 replied:
Thank you @anselld

hansieodendaal replied:
Hi @Empiech007, it has been updated. Maybe just a problem with what you see.

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
Please remove all back-ticks (`) in this text block.

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
````diff
@@ -735,508 +505,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
+**Algorithm 1**: This algorithm computes new distinct doubles as soon as it is possible to do so. These new doubles are 
+turned into triples in the next immediate IPP round. This process is repeated until all IPP rounds are completed. Only 
+then are next larger-sized sub-products computed, but consumption of the smallest existing "tuples" is given priority. 
+```text
````
Please remove all back-ticks (`) inside this text block, between the opening (```Text) and closing (```).

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
````diff
@@ -735,508 +505,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
+**Algorithm 2**: This algorithm starts exactly like Algorithm 1, by forming only new and distinct doubles and turning 
+them into triples in the next immediate IPP round, but goes beyond the triples by immediately forming the next possible 
+quadruples.
+```text
````
Please remove all back-ticks (`) inside this text block, between the opening (```Text) and closing (```).

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
````diff
@@ -735,508 +505,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
+**Algorithm 3**: This algorithm computes new distinct doubles as soon as it is possible to do so, until the end of the 
+IPP rounds. The program then forms any possible triples. Larger-sized sub-products are then computed by firstly 
+consuming the smallest existing "tuples". 
+```text
````
Please remove all back-ticks (`) inside this text block, between the opening (```Text) and closing (```).

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
````diff
@@ -735,508 +505,316 @@ The following table displays the multiplication cost of the Naive Algorithm toge
+**Algorithm 4**: This algorithm is the same as Algorithm 3 throughout the IPP rounds. However, at the end of the IPP 
+rounds, the program gives preference to the formation of all possible distinct quadruples. Larger-sized sub-products are then 
+computed by firstly consuming the largest existing "tuples". 
+```text
````
Please remove all back-ticks (`) inside this text block, between the opening (```Text) and closing (```).

hansieodendaal commented on src/cryptography/amortization-bp-ipp/mainreport.md:
This can be added back into the text block, similar to the other algorithms.

anselld commented:
> Is this intentional?
> ![image](https://user-images.githubusercontent.com/39146854/82429683-9d194780-9a8c-11ea-895b-6820333de14e.png)

Empiech007 commented on src/cryptography/amortization-bp-ipp/mainreport.md:
```diff
@@ -1,729 +1,499 @@
+$\big( \mathbf{a}^{(j)} , \mathbf{b}^{(j)} , \mathbf{G}^{(j)} , \mathbf{H}^{(j)} \big)​$, which is initially 
+$\big( \mathbf{a} , \mathbf{b} , \mathbf{G} , \mathbf{H} \big) ​$. However, when $j < k​$, the input is updated to 
+$\big(\mathbf{a}^{(j-1)}, \mathbf{b}^{(j-1)}, \mathbf{G}^{(j-1)},\mathbf{H}^{(j-1)} \big)​$ quadruple vectors each of 
+size $2^{k-1}​$, where $\mathbf{a}^{(j-1)} = \mathbf{a}\_{lo} \cdot u\_j + \mathbf{a}\_{hi} \cdot u\_j^{-1} ​$, \ \ 
+$\mathbf{b}^{(j-1)} = \mathbf{b}\_{lo} \cdot u\_j^{-1} + \mathbf{b}\_{hi} \cdot u\_j​$, \ \ 
+$\mathbf{G}^{(j-1)} = \mathbf{G}\_{lo} \cdot u\_j^{-1} + \mathbf{G}\_{hi} \cdot u\_j​$,\ \ 
+$\mathbf{H}^{(j-1)} = \mathbf{H}\_{lo} \cdot u\_j  + \mathbf{H}\_{hi} \cdot u\_j^{-1}​$, and $u_k​$ is the verifier's 
+challenge. The vectors\ \ $\mathbf{a}\_{lo}, \mathbf{b}\_{lo}, \mathbf{G}\_{lo}, \mathbf{H}\_{lo}​$\ \ and \ \ 
+$\mathbf{a}\_{hi}, \mathbf{b}\_{hi}, \mathbf{G}_{hi}, \mathbf{H}\_{hi} ​$\ \ 
+being the left and the right halves of the vectors \ \ $\mathbf{a}, \mathbf{b}, \mathbf{G}, \mathbf{H}​$, respectively. 
```
Yes, the inserted " \ \ " were intentional, however if they create wide spaces unnecessarily then they can be removed.

hansieodendaal replied:
This has been fixed

anselld commented:
Hi @Empiech007, please remember to change 'Initialise' to 'Initialize' in all the figures, for consistency with the text.
Please let me know if you'd like any other changes. Thanks.

Empiech007 commented:
Yes, @anselld , I have the corrected picture ready for once you have finalised editing. 

Please check my most recent comments. Thank you

hansieodendaal commented:
Thank you @anselld and @Empiech007, I made the final presentation layer fixes. @Empiech007, you can edit your image. @anselld, please finally verify all links. (_Remember to pull the latest commit,_)

anselld commented:
Thanks for dividing the appendices into A and B, @hansieodendaal. I re-checked all the links as requested - all are okay. Please remember, @Empiech007, to include the updated figures. Thanks!

Empiech007 commented:
Figure updated to 'recursive-zk-proofs' ... @anselld, @hansieodendaal

Empiech007 commented:
There is **only one test**, hence the original subheading "Verifier's Test of Prover's  $s_i$  values" 
... let's rather have
 
 - **Verifier's Test of Prover's  $s_i$  Values** 
    - **Properties of the  $s_i$  Values**  
    - **Test of the $s_i$ Values** 

Thanks @anselld and @hansieodendaal ... great job.

hansieodendaal commented:
Hi @Empiech007, there were merge conflicts with the base branch. Please redo your changes in this branch and commit again to this branch. Make sure you pull the latest. Thanks.

Empiech007 commented:
I commited and pushed again to 'recursive zk proofs'.

hansieodendaal commented:
Hi @Empiech007, you have to apply your changes to this branch as it is temporarily the master: `recursive-zk-proofs-da-edits`. Then Denise can also verify your changes.

Empiech007 commented:
I don't have access to 'Recursive zk proofs da edits #257' via GitKraken, it does not appear on my GitKraken.

hansieodendaal commented:
@Empiech007, figure updated in this branch. 

**FYI:** In GitKraken, checkout and pull master, then check out this branch.
![image](https://user-images.githubusercontent.com/39146854/82652763-74c15280-9c1e-11ea-84d4-ed86b730a05e.png)   ![image](https://user-images.githubusercontent.com/39146854/82652865-9faba680-9c1e-11ea-80c0-3fa31c3c6466.png)

hansieodendaal commented:
> There is **only one test**, hence the original subheading "Verifier's Test of Prover's $s_i$ values"
> ... let's rather have
> 
> * **Verifier's Test of Prover's  $s_i$  Values**
>   
>   * **Properties of the  $s_i$  Values**
>   * **Test of the $s_i$ Values**
> 
> Thanks @anselld and @hansieodendaal ... great job.

@anselld and @Empiech007, please fix this.

hansieodendaal commented:
@anselld, please see the changes I made to making hyperlinks for figures and tables in [commit 5bb7217](https://github.com/tari-labs/tari-university/pull/257/commits/5bb7217ad2c090f5b18eaa4c632987b2315d2b34), for future reference.

Empiech007 commented:
"FYI: In GitKraken, checkout and pull master, then check out this branch." 

I have done all that, but I still don’t have the 'recursive-zk-proofs-da-edits'

anselld commented:
Hi @hansieodendaal, thanks for the comment about making hyperlinks for figures and tables. I see that the names of the figures and tables in the code were changed. Do you just use the initial letter of the caption for the figures and tables?

hansieodendaal commented:
> "FYI: In GitKraken, checkout and pull master, then check out this branch." @hansieodendaal
> 
> I have done all that, but I still don’t have the 'recursive-zk-proofs-da-edits'

Then you have to use git on the command line, the branch exists, you can check it out.

Empiech007 commented:
Okay, it is done. Headings fixed and Picture (Figure 2) updated. 
Fantastic ... Thanks @hansieodendaal and @anselld

anselld commented:
@hansieodendaal and @Empiech007, I've updated the headings
Verifier's Test of Prover's $s_i$ Values
Properties of the $s_i$ Values
Test of the $s_i$ Values
I've also corrected the contents list accordingly, and have checked these links.

Empiech007 commented:
For consistency I have changed all "$s_i$ Values" to "$G_i$s Coefficients" in headings. 
I checked the links but please double-check @anselld ... Thanks

hansieodendaal commented:
> @hansieodendaal and @Empiech007, I've updated the headings
> Verifier's Test of Prover's $s_i$ Values
> Properties of the $s_i$ Values
> Test of the $s_i$ Values
> I've also corrected the contents list accordingly, and have checked these links.

Hi @anselld, I cannot see your commit. Have you pushed? Just make sure you pull first because @Empiech007 pushed 2x commits. Thanks

anselld commented:
Hi @Empiech007, final run-through of links has been done, latest commit has been pushed to GitHub.

hansieodendaal commented:
Hi @anselld  and @Empiech007, freak branches crated on Friday found and corrected, all good now. Latest changes made by @Empiech007 now in branch. Merging. Thanks!
